File: .agents/skills/codebase-security/scripts/typescript/vuln_deps.js
#!/usr/bin/env node
/**
 * Layer 5 (TypeScript/JavaScript) — known-vulnerable dependencies (SCA).
 *
 * Picks the audit command from the lockfile present:
 *   package-lock.json → npm audit --json
 *   pnpm-lock.yaml    → pnpm audit --json
 *   yarn.lock         → yarn audit --json (classic, ndjson)
 *
 * osv-scanner (if on PATH) runs additionally and is merged in — it is the
 * only one of these that flags OSV `MAL-` malicious-package advisories.
 * All of them need network access; offline → skipped with warning.
 */
'use strict';

const { spawnSync } = require('node:child_process');
const fs = require('node:fs');
const path = require('node:path');

const SEV_ORDER = { critical: 4, high: 3, moderate: 2, medium: 2, low: 1, info: 0 };

function normSev(s) {
  s = String(s || 'unknown').toLowerCase();
  return s === 'moderate' ? 'medium' : (SEV_ORDER[s] !== undefined ? s : 'unknown');
}

function run(cmd, args, cwd) {
  return spawnSync(cmd, args, {
    cwd, encoding: 'utf8', maxBuffer: 64 * 1024 * 1024, timeout: 600_000,
  });
}

function npmAudit(repo) {
  const res = run('npm', ['audit', '--json'], repo);
  if (!res.stdout || !res.stdout.trim().startsWith('{')) return null;
  let data;
  try { data = JSON.parse(res.stdout); } catch { return null; }
  const vulns = data.vulnerabilities || {};
  const findings = [];
  for (const [name, v] of Object.entries(vulns)) {
    const advisories = (v.via || []).filter((x) => typeof x === 'object');
    findings.push({
      package: name,
      version: v.range || '?',
      vuln_id: advisories[0] ? (advisories[0].url || advisories[0].title || '?') : 'transitive',
      aliases: advisories.map((a) => a.cve || a.title).filter(Boolean).slice(0, 5),
      severity: normSev(v.severity),
      malicious: false,
      summary: advisories[0] ? String(advisories[0].title || '').slice(0, 200) : `via ${(v.via || []).filter((x) => typeof x === 'string').join(', ')}`,
      fixed_in: v.fixAvailable ? [typeof v.fixAvailable === 'object' ? `${v.fixAvailable.name}@${v.fixAvailable.version}` : 'run npm audit fix'] : [],
      direct: !!v.isDirect,
      manifest: 'package-lock.json',
      tool: 'npm-audit',
    });
  }
  return findings;
}

function pnpmAudit(repo) {
  const res = run('pnpm', ['audit', '--json'], repo);
  if (!res.stdout || !res.stdout.trim().startsWith('{')) return null;
  let data;
  try { data = JSON.parse(res.stdout); } catch { return null; }
  const findings = [];
  for (const adv of Object.values(data.advisories || {})) {
    findings.push({
      package: adv.module_name,
      version: adv.vulnerable_versions || '?',
      vuln_id: (adv.cves || [])[0] || `GHSA:${adv.github_advisory_id || adv.id}`,
      aliases: adv.cves || [],
      severity: normSev(adv.severity),
      malicious: false,
      summary: String(adv.title || '').slice(0, 200),
      fixed_in: adv.patched_versions ? [adv.patched_versions] : [],
      direct: false,
      manifest: 'pnpm-lock.yaml',
      tool: 'pnpm-audit',
    });
  }
  return findings;
}

function yarnAudit(repo) {
  const res = run('yarn', ['audit', '--json'], repo);
  if (!res.stdout) return null;
  const findings = [];
  for (const line of res.stdout.split('\n')) {
    if (!line.trim().startsWith('{')) continue;
    let row;
    try { row = JSON.parse(line); } catch { continue; }
    if (row.type !== 'auditAdvisory') continue;
    const adv = row.data.advisory;
    findings.push({
      package: adv.module_name,
      version: adv.vulnerable_versions || '?',
      vuln_id: (adv.cves || [])[0] || `advisory:${adv.id}`,
      aliases: adv.cves || [],
      severity: normSev(adv.severity),
      malicious: false,
      summary: String(adv.title || '').slice(0, 200),
      fixed_in: adv.patched_versions ? [adv.patched_versions] : [],
      direct: false,
      manifest: 'yarn.lock',
      tool: 'yarn-audit',
    });
  }
  return findings;
}

function osvScanner(repo) {
  const probe = spawnSync('osv-scanner', ['--version'], { encoding: 'utf8' });
  if (probe.error) return null;
  for (const args of [
    ['scan', 'source', '-r', '--format', 'json', repo],
    ['--recursive', '--format', 'json', repo],
  ]) {
    const res = run('osv-scanner', args, repo);
    if (!res.stdout || !res.stdout.trim().startsWith('{')) continue;
    let data;
    try { data = JSON.parse(res.stdout); } catch { continue; }
    const findings = [];
    for (const r of data.results || []) {
      for (const pkg of r.packages || []) {
        const info = pkg.package || {};
        if (info.ecosystem !== 'npm') continue;
        for (const vuln of pkg.vulnerabilities || []) {
          const malicious = String(vuln.id || '').startsWith('MAL-') ||
            (vuln.aliases || []).some((a) => String(a).startsWith('MAL-'));
          const dbSev = ((vuln.database_specific || {}).severity || '').toLowerCase();
          findings.push({
            package: info.name,
            version: info.version,
            vuln_id: vuln.id,
            aliases: vuln.aliases || [],
            severity: malicious ? 'critical' : normSev(dbSev),
            malicious,
            summary: String(vuln.summary || '').slice(0, 200),
            fixed_in: [],
            direct: false,
            manifest: (r.source || {}).path || '?',
            tool: 'osv-scanner',
          });
        }
      }
    }
    return findings;
  }
  return null;
}

function main() {
  const args = process.argv.slice(2);
  const repo = path.resolve(args[0]);
  const out = args[args.indexOf('--out') + 1];
  fs.mkdirSync(path.dirname(out), { recursive: true });

  const warnings = [];
  let findings = null;
  let tool = 'none';

  if (fs.existsSync(path.join(repo, 'package-lock.json'))) {
    findings = npmAudit(repo); tool = 'npm-audit';
  } else if (fs.existsSync(path.join(repo, 'pnpm-lock.yaml'))) {
    findings = pnpmAudit(repo); tool = 'pnpm-audit';
  } else if (fs.existsSync(path.join(repo, 'yarn.lock'))) {
    findings = yarnAudit(repo); tool = 'yarn-audit';
  } else {
    warnings.push('no lockfile found — package-manager audit cannot run (also flagged by supply_chain layer)');
  }
  if (findings === null && tool !== 'none') {
    warnings.push(`${tool} failed (offline? registry unreachable?)`);
    findings = null;
  }

  // Merge osv-scanner results (dedupe on package+vuln_id).
  const osv = osvScanner(repo);
  if (osv) {
    const seen = new Set((findings || []).map((f) => `${f.package}|${f.vuln_id}`));
    findings = (findings || []).concat(osv.filter((f) => !seen.has(`${f.package}|${f.vuln_id}`)));
    tool = tool === 'none' ? 'osv-scanner' : `${tool}+osv-scanner`;
  }

  let payload;
  if (findings === null) {
    payload = {
      layer: 'vuln_deps', skipped: true,
      reason: warnings.join('; ') || 'no SCA tool succeeded',
      findings: [], counts: { total: 0 }, warnings,
    };
    console.log(`vuln_deps(ts): SKIPPED — ${payload.reason}`);
  } else {
    const bySev = {};
    for (const s of ['critical', 'high', 'medium', 'low', 'unknown']) {
      bySev[s] = findings.filter((f) => f.severity === s).length;
    }
    payload = {
      layer: 'vuln_deps',
      tool,
      findings,
      counts: {
        total: findings.length,
        malicious: findings.filter((f) => f.malicious).length,
        fixable: findings.filter((f) => f.fixed_in.length).length,
        by_severity: bySev,
      },
      warnings,
    };
    console.log(`vuln_deps(ts): ${findings.length} vulns via ${tool} (critical=${bySev.critical}, high=${bySev.high})`);
  }
  fs.writeFileSync(out, JSON.stringify(payload, null, 2));
}

main();

File: .agents/skills/codebase-security/scripts/python/vuln_deps.py
#!/usr/bin/env python3
"""Layer 5 (Python) — known-vulnerable dependencies (SCA).

Strategy, in order of preference:
1. osv-scanner on the repo (covers uv.lock, poetry.lock, requirements.txt,
   Pipfile.lock — and flags OSV `MAL-` malicious-package advisories).
2. pip-audit (PATH or `uvx pip-audit`) against requirements.txt, or against
   a temporary export of uv.lock (`uv export --format requirements-txt`).

Both need network access (OSV / PyPI advisory APIs). Offline → layer is
skipped with an explicit warning, never silently green.
"""
from __future__ import annotations

import argparse
import json
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path


def severity_from_osv(vuln: dict) -> str:
    """Normalise the GHSA-style severity OSV carries in database_specific."""
    ds = (vuln.get("database_specific") or {}).get("severity", "")
    return {"CRITICAL": "critical", "HIGH": "high", "MODERATE": "medium",
            "MEDIUM": "medium", "LOW": "low"}.get(ds.upper(), "unknown")


def run_osv_scanner(repo: Path) -> tuple[list[dict] | None, str | None]:
    if not shutil.which("osv-scanner"):
        return None, None
    # v2 dialect first, then v1.
    for args in (
        ["osv-scanner", "scan", "source", "-r", "--format", "json", str(repo)],
        ["osv-scanner", "--recursive", "--format", "json", str(repo)],
    ):
        try:
            result = subprocess.run(args, capture_output=True, text=True, check=False, timeout=600)
        except (FileNotFoundError, subprocess.TimeoutExpired):
            return None, "osv-scanner timed out or vanished"
        # exit 1 = vulns found (still valid JSON); 127/128 = bad CLI dialect
        if result.stdout.strip().startswith("{"):
            try:
                data = json.loads(result.stdout)
            except json.JSONDecodeError:
                continue
            findings = []
            for res in data.get("results", []):
                source = res.get("source", {}).get("path", "")
                for pkg in res.get("packages", []):
                    info = pkg.get("package", {})
                    if info.get("ecosystem") not in ("PyPI", ""):
                        continue
                    for vuln in pkg.get("vulnerabilities", []):
                        vid = vuln.get("id", "?")
                        aliases = vuln.get("aliases", []) or []
                        is_malicious = vid.startswith("MAL-") or any(a.startswith("MAL-") for a in aliases)
                        findings.append({
                            "package": info.get("name", "?"),
                            "version": info.get("version", "?"),
                            "vuln_id": vid,
                            "aliases": aliases,
                            "severity": "critical" if is_malicious else severity_from_osv(vuln),
                            "malicious": is_malicious,
                            "summary": (vuln.get("summary") or "")[:200],
                            "fixed_in": _osv_fixed_versions(vuln, info.get("name", "")),
                            "manifest": source,
                            "tool": "osv-scanner",
                        })
            return findings, None
    return None, "osv-scanner present but both CLI dialects failed"


def _osv_fixed_versions(vuln: dict, pkg: str) -> list[str]:
    fixed = []
    for aff in vuln.get("affected", []) or []:
        if aff.get("package", {}).get("name", "").lower() != pkg.lower():
            continue
        for r in aff.get("ranges", []) or []:
            for ev in r.get("events", []) or []:
                if "fixed" in ev:
                    fixed.append(ev["fixed"])
    return sorted(set(fixed))[:3]


def pip_audit_cmd() -> list[str] | None:
    if shutil.which("pip-audit"):
        return ["pip-audit"]
    if shutil.which("uvx"):
        return ["uvx", "pip-audit"]
    return None


def requirements_source(repo: Path) -> tuple[Path | None, bool, str | None]:
    """Return (requirements file, is_temporary, warning)."""
    req = repo / "requirements.txt"
    if req.exists():
        return req, False, None
    if (repo / "uv.lock").exists() and shutil.which("uv"):
        tmp = Path(tempfile.mkstemp(suffix="-requirements.txt")[1])
        result = subprocess.run(
            ["uv", "export", "--format", "requirements-txt", "--no-emit-project",
             "--no-hashes", "-o", str(tmp)],
            cwd=str(repo), capture_output=True, text=True, check=False,
        )
        if result.returncode == 0:
            return tmp, True, None
        tmp.unlink(missing_ok=True)
        return None, False, f"uv export failed: {result.stderr[:200]}"
    return None, False, "no requirements.txt and no uv.lock+uv to export from"


def run_pip_audit(repo: Path) -> tuple[list[dict] | None, str | None]:
    cmd = pip_audit_cmd()
    if cmd is None:
        return None, "pip-audit not available (install pip-audit, or uv for uvx fallback)"
    req, temporary, warn = requirements_source(repo)
    if req is None:
        return None, warn
    try:
        # --no-deps --disable-pip works on fully pinned files (uv/poetry
        # exports always are) without building a venv. Unpinned hand-written
        # requirements need the slower venv-based resolution — retry without
        # the fast-path flags if pinning is the complaint.
        base = cmd + ["-r", str(req), "--format", "json", "--progress-spinner", "off"]
        result = subprocess.run(
            base + ["--no-deps", "--disable-pip"],
            capture_output=True, text=True, check=False, timeout=600,
        )
        if result.returncode and "pinned" in (result.stderr or "").lower():
            result = subprocess.run(base, capture_output=True, text=True,
                                    check=False, timeout=600)
    except subprocess.TimeoutExpired:
        return None, "pip-audit timed out (network?)"
    finally:
        if temporary:
            req.unlink(missing_ok=True)
    if not result.stdout.strip().startswith("{") and not result.stdout.strip().startswith("["):
        return None, f"pip-audit failed: {(result.stderr or result.stdout)[:300]}"
    try:
        data = json.loads(result.stdout)
    except json.JSONDecodeError:
        return None, "pip-audit emitted invalid JSON"

    deps = data.get("dependencies", data if isinstance(data, list) else [])
    findings = []
    for dep in deps:
        for v in dep.get("vulns", []):
            findings.append({
                "package": dep.get("name", "?"),
                "version": dep.get("version", "?"),
                "vuln_id": v.get("id", "?"),
                "aliases": v.get("aliases", []),
                "severity": "unknown",  # pip-audit JSON has no severity field
                "malicious": False,
                "summary": (v.get("description") or "")[:200],
                "fixed_in": v.get("fix_versions", []),
                "manifest": "requirements.txt|uv.lock",
                "tool": "pip-audit",
            })
    return findings, None


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("repo")
    ap.add_argument("--targets", required=False)  # unused; manifest-level layer
    ap.add_argument("--out", required=True)
    args = ap.parse_args()

    repo = Path(args.repo).resolve()
    out = Path(args.out)
    out.parent.mkdir(parents=True, exist_ok=True)

    warnings: list[str] = []
    findings, warn = run_osv_scanner(repo)
    tool = "osv-scanner"
    if warn:
        warnings.append(warn)
    if findings is None:
        findings, warn = run_pip_audit(repo)
        tool = "pip-audit"
        if warn:
            warnings.append(warn)

    if findings is None:
        payload = {
            "layer": "vuln_deps", "skipped": True,
            "reason": "; ".join(warnings) or "no SCA tool available",
            "findings": [], "counts": {"total": 0}, "warnings": warnings,
        }
        out.write_text(json.dumps(payload, indent=2))
        print(f"vuln_deps(py): SKIPPED — {payload['reason']}")
        return 0

    by_sev = {s: sum(1 for f in findings if f["severity"] == s)
              for s in ("critical", "high", "medium", "low", "unknown")}
    payload = {
        "layer": "vuln_deps",
        "tool": tool,
        "findings": findings,
        "counts": {
            "total": len(findings),
            "malicious": sum(1 for f in findings if f["malicious"]),
            "fixable": sum(1 for f in findings if f["fixed_in"]),
            "by_severity": by_sev,
        },
        "warnings": warnings,
    }
    out.write_text(json.dumps(payload, indent=2))
    print(f"vuln_deps(py): {len(findings)} vulns via {tool} "
          f"(critical={by_sev['critical']}, high={by_sev['high']})")
    return 0


if __name__ == "__main__":
    sys.exit(main())
